controllers/etcdpeer_controller.go

```diff
@@ -16,20 +21,115 @@ type EtcdPeerReconciler struct {
 	Log logr.Logger
 }
 
+const (
+	etcdImage = "quay.io/coreos/etcd:v3.2.27"
+)
+
 // +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=replicaset,verbs=get;update;patch;create
+
+func defineReplicaSet(etcdPeer etcdv1alpha1.EtcdPeer) (appsv1.ReplicaSet, error) {
+	var replicas int32 = 1
+
+	return appsv1.ReplicaSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:          make(map[string]string),
+			Annotations:     make(map[string]string),
+			Name:            etcdPeer.Name,
+			Namespace:       etcdPeer.Namespace,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(&etcdPeer, etcdv1alpha1.GroupVersion.WithKind("EtcdPeer"))},
+		},
+		Spec: appsv1.ReplicaSetSpec{
+			// This will *always* be 1. Other peers are handled by other EtcdPeers.
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": "etcd",
+					// Using the EtcdPeer's name as a label limits what the name can be
+					"peer": etcdPeer.Name,
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": "etcd",
+						// Using the EtcdPeer's name as a label limits what the name can be
+						"peer": etcdPeer.Name,
+					},
+					Annotations: make(map[string]string),
+					Name:        etcdPeer.Name,
+					Namespace:   etcdPeer.Namespace,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "etcd",
+							Image: etcdImage,
+						},
+					},
+				},
+			},
+		},
+	}, nil
+}
 
 func (r *EtcdPeerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("etcdpeer", req.NamespacedName)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	log := r.Log.WithValues("etcdpeer", req.NamespacedName)
+
+	var etcdPeer etcdv1alpha1.EtcdPeer
+	if err := r.Get(ctx, req.NamespacedName, &etcdPeer); err != nil {
+		log.Error(err, "unable to fetch EtcdPeer")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.V(2).Info("Found EtcdPeer", "name", etcdPeer.Name)
+
+	var existingReplicaSet appsv1.ReplicaSet
+	err := r.Get(
+		ctx,
+		client.ObjectKey{
+			Namespace: etcdPeer.Namespace,
+			Name:      etcdPeer.Name,
+		},
+		&existingReplicaSet,
+	)
+
+	if apierrs.IsNotFound(err) {
+		log.V(1).Info("Replica set does not exist, creating")
+		replicaSet, err := defineReplicaSet(etcdPeer)
+		if err != nil {
+			log.Error(err, "unable to generate ReplicaSet from EtcdPeer")
+			return ctrl.Result{}, err
```

Comment:
Not that it's an issue right now because `defineReplicaSet` doesn't ever return an error, but when it comes to extending these functions out, you should take special note of whether we *do* return an error here.

Always remember that returning an error will cause the Reconcile function to be called again automatically. The question you should ask yourself is "is retrying this operation likely to cause it to succeed next time, or will it require some other configuration/state of the world to change?"

I don't think it should block merging it now, but it's definitely something to bare in mind 😄

Reply:
I had `defineReplicaSet` return an error as it seemed cleaner, but as it can never fail I'm wondering if I should just apply YAGNI and make it not return an error at all.

Reply:
> Always remember that returning an error will cause the Reconcile function to be called again automatically.

How do I then express "there is an error, but a retry won't fix it"?

Reply:
Speculative question, so moving this elsewhere.